qadence/ml_tools/callbacks/writer_registry.py

```diff
@@ -0,0 +1,375 @@
+from __future__ import annotations
+
+import os
+from logging import getLogger
+from types import ModuleType
+from typing import Any, Callable, Union
+from uuid import uuid4
+
+from matplotlib.figure import Figure
+from mlflow.entities import Run
+from torch import Tensor
+from torch.nn import Module
+from torch.utils.data import DataLoader
+from torch.utils.tensorboard import SummaryWriter
+
+from qadence.ml_tools.config import TrainConfig
+from qadence.ml_tools.data import DictDataLoader, OptimizeResult
+from qadence.types import ExperimentTrackingTool
+
+logger = getLogger(__name__)
+
+# Type aliases
+PlottingFunction = Callable[[Module, int], tuple[str, Figure]]
+InputData = Union[Tensor, dict[str, Tensor]]
+
+
+class BaseWriter:
+    """
+    Abstract base class for experiment tracking writers.
+
+    Methods:
+        open(config, iteration=None): Opens the writer and sets up the logging
+            environment.
+        close(): Closes the writer and finalizes any ongoing logging processes.
+        print_metrics(result): Prints metrics and loss in a formatted manner.
+        write(result): Writes the optimization results to the tracking tool.
+        log_hyperparams(hyperparams): Logs the hyperparameters to the tracking tool.
+        plot(model, iteration, plotting_functions): Logs model plots using provided
+            plotting functions.
+        log_model(model, dataloader): Logs the model and any relevant information.
+    """
+
+    run: Run  # [attr-defined]
+
+    def open(self, config: TrainConfig, iteration: int = None) -> Any:
```

Comment:
Shouldn't these methods be `@abstractmethod`s ?

Reply:
Updated.